src/script/lua_api/l_item.cpp

```diff
@@ -29,6 +29,16 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 #include "log.h"
 
 
+// ItemStack metamethod
+// __eq(item1, item2) -> true/false
+int LuaItemStack::itemstack_compare(lua_State *L)
+{
+	NO_MAP_LOCK_REQUIRED;
+	lua_pushboolean(L, checkobject(L, 1)->m_stack == checkobject(L, 2)->m_stack);
```

Comment:
this may cause a segfault if comparing incompatible types

Reply:
If checkobject fails it throws a Lua error, so how could it segfault?
Also I think the __eq metamethod will only be called when both values have the same type.

Reply:
Ahh, of course

>  Also I think the __eq metamethod will only be called when both values have the same type.

This is not true, they can have any type and it's not even guaranteed that the first argument is the correct type

Reply:
> If checkobject fails it throws a Lua error, so how could it segfault?
Shouldn't it return false in that case?

Reply:
The last paragraph of https://www.lua.org/pil/13.2.html says:

>An equality comparison never raises an error, but if two objects have different metamethods, the equality operation results in false, without even calling any metamethod. Again, this behavior mimics the common behavior of Lua, which always classifies strings as different from numbers, regardless of their values. Lua calls the equality metamethod only when the two objects being compared share this metamethod.

So __eq will always be called with itemstacks unless you could directly call the __eq function.

Reply:
Oh, I'm probably thinking of __concat